frappe/tests/test_translate.py

```diff
@@ -125,6 +133,91 @@ def test_load_all_translate_files(self):
 		"""Load all CSV files to ensure they have correct format"""
 		verify_translation_files("frappe")
 
+	def test_python_extractor(self):
+
+		code = textwrap.dedent(
+			"""
+			frappe._("attr")
+			_("name")
+			frappe._("attr with", context="attr context")
+			_("name with", context="name context")
+			_("broken on",
+				context="new line")
+			__("This wont be captured")
+			__init__("This shouldn't too")
+			_(
+				"broken on separate line",
+				)
+			_(not_a_string)
+			_(not_a_string, context="wat")
+		"""
+		)
+		expected_output = [
+			(2, "attr", None),
+			(3, "name", None),
+			(4, "attr with", "attr context"),
+			(5, "name with", "name context"),
+			(6, "broken on", "new line"),
+			(10, "broken on separate line", None),
+		]
+
+		output = extract_messages_from_python_code(code)
+		self.assertEqual(len(expected_output), len(output))
+		for expected, actual in zip(expected_output, output):
+			with self.subTest():
+				self.assertEqual(expected, actual)
+
+	def test_js_extractor(self):
+
+		code = textwrap.dedent(
+			"""
+			__("attr")
+			__("attr with", null, "context")
+			__("attr with", ["format", "replacements"], "context")
+			__("attr with", ["format", "replacements"])
+			__(
+				"Long JS string with", [
+					"format", "replacements"
+				],
+				"JS context on newline"
+			)
+			__(
+				"Long JS string with formats only {0}", [
+					"format", "replacements"
+				],
+			)
+			_(`template strings not supported yet`)
```

Comment:
Second underscore missing here

Reply:
Ah, the message inside test string is misleading, it's meant to test that `_` doesn't get captured. Previous regex didn't respect the actual translation function for the language type. 

Equivalent python version:
https://github.com/frappe/frappe/blob/f6210e7c38d33f99f54d22067228b0f9ec558d4d/frappe/tests/test_translate.py#L146



Template strings still won't work if they have replacement inside them, else they will work.... similar to python